packages/password-input/src/PasswordInput/PasswordInput.tsx

```diff
@@ -0,0 +1,208 @@
+import React, { useEffect, useState } from 'react';
+import PropTypes from 'prop-types';
+
+import { cx } from '@leafygreen-ui/emotion';
+import { useControlledValue, useIdAllocator } from '@leafygreen-ui/hooks';
+import LeafyGreenProvider, {
+  useDarkMode,
+} from '@leafygreen-ui/leafygreen-provider';
+import { Label } from '@leafygreen-ui/typography';
+
+import { InputIcon } from '../InputIcon';
+import { StateNotifications } from '../StateNotifications';
+import { TogglePassword } from '../TogglePassword';
+
+import {
+  inputBaseStyles,
+  inputBaseThemeStyles,
+  inputDisabledBaseStyles,
+  inputDisabledThemeStyles,
+  inputIconSizeStyles,
+  inputSizeStyles,
+  inputThemeStyles,
+  inputWrapperStyles,
+  labelBaseStyles,
+  labelLargeOverrideStyles,
+} from './PasswordInput.styles';
+import {
+  type NotificationProps,
+  PasswordInputProps,
+  SizeVariant,
+  States,
+} from './PasswordInput.types';
+import { getStateFromArray } from './utils';
+
+export const PasswordInput = React.forwardRef<
+  HTMLInputElement,
+  PasswordInputProps
+>(
+  (
+    {
+      value: valueProp,
+      onChange: onChangeProp,
+      className,
+      darkMode: darkModeProp,
+      label,
+      sizeVariant = SizeVariant.Default,
+      stateNotifications = [],
+      disabled = false,
+      autoComplete = 'new-password',
+      id: idProp,
+      'aria-describedby': ariaDescribedbyProp,
+      'aria-labelledby': ariaLabelledbyProp,
+      ...rest
+    }: PasswordInputProps,
+    forwardedRef,
+  ) => {
+    const [showPassword, setShowPassword] = useState<boolean>(false);
+    const prefix = 'passwordinput';
+    const inputId = useIdAllocator({ prefix, id: idProp });
+    const ariaDescribedby = useIdAllocator({
+      prefix,
+      id: ariaDescribedbyProp,
+    });
+    const ariaLabelledby = useIdAllocator({
+      prefix,
+      id: ariaLabelledbyProp,
+    });
+    const { theme, darkMode } = useDarkMode(darkModeProp);
+    const { value, handleChange } = useControlledValue(valueProp, onChangeProp);
+
+    if (!label && !ariaLabelledbyProp && !rest['aria-label']) {
+      console.warn(
+        'For screen-reader accessibility, label, aria-labelledby, or aria-label must be provided',
+      );
+    }
+
+    // If disabled then hide password
+    useEffect(() => {
+      if (disabled) setShowPassword(false);
+    }, [disabled]);
+
+    const handleTogglePasswordClick = () => setShowPassword(s => !s);
+
+    /**
+     * The overall state of the component
+     */
+    const state: States = Array.isArray(stateNotifications)
+      ? getStateFromArray(stateNotifications)
+      : stateNotifications;
+
+    /**
+     * Checks if there are any notifications
+     */
+    const hasNotifications: boolean =
+      !ariaDescribedbyProp && Array.isArray(stateNotifications);
```

Comment:
I think you'll want to check the array length rather than using isArray here, since you've set the prop to default to an empty array

Reply:
Since I want the message container to render even if there are no messages I think this should be ok. The only time it should not render is if a custom message container is used.